lib/cli-engine/cli-engine.js

```diff
@@ -804,24 +875,110 @@ class CLIEngine {
             if (lintResultCache) {
                 lintResultCache.setCachedLintResults(filePath, config, result);
             }
-        }
+            return { result, config };
+        });
 
-        // Persist the cache to disk.
-        if (lintResultCache) {
-            lintResultCache.reconcile();
-        }
+        const results = verifyFilesResults.map(({ result }) => result);
+        const lastConfigArraysFromResults = verifyFilesResults.map(({ config }) => config).filter(item => item);
 
-        // Collect used deprecated rules.
-        const usedDeprecatedRules = Array.from(
-            iterateRuleDeprecationWarnings(lastConfigArrays)
-        );
+        return { results, lastConfigArraysFromResults };
+    }
 
-        debug(`Linting complete in: ${Date.now() - startTime}ms`);
-        return {
-            results,
-            ...calculateStatsPerRun(results),
-            usedDeprecatedRules
-        };
+    async _verifyFilesAsync(files) {
+        const {
+            options: {
+                parallel
+            }
+        } = internalSlotsMap.get(this);
+
+        const workersCount = parallel;
+
+        const pool = workerpool.pool(`${__dirname}/worker.js`, {
+            maxWorkers: workersCount,
+            workerType: "process",
+            minWorkers: "max"
+        });
+
+        /*
+         * send files without config to workers to avoid serializing config
+         * config will be loaded on the workers
+         */
+        const filesWithoutConfig = files.map(({ filePath, ignored }) => ({ filePath, ignored }));
+
+        const chunkSize = Math.ceil(filesWithoutConfig.length / workersCount);
+        const chunks = lodash.chunk(filesWithoutConfig, chunkSize);
+
+        const chunkResults = await Promise.all(chunks.map(async(chunkFiles, chunkIndex) =>
+            await pool.exec("lintChunkOnWorker", [{
+                params: { chunkFiles, chunkIndex },
+                providedOptions: this.providedOptions
+            }])));
+
+        pool.terminate();
+        const verifyFilesResults = [];
+
+        chunkResults.forEach(chunkResult => {
+            verifyFilesResults.push(...chunkResult);
+        });
+
+        const results = verifyFilesResults.map(({ result }) => result);
+        const lastConfigArraysFromResults = verifyFilesResults.map(({ config }) => config).filter(item => item);
+
+        return { results, lastConfigArraysFromResults };
+    }
+
+    async lintChunkOnWorker({ chunkFiles }) {
+        const {
+            fileEnumerator,
+            lintResultCache,
+            linter,
+            configArrayFactory,
+            options: {
+                allowInlineConfig,
+                cwd,
+                fix,
+                reportUnusedDisableDirectives
+            }
+        } = internalSlotsMap.get(this);
+
+        const verifyFilesResults = await bluebird.map(chunkFiles, async({ filePath, ignored }) => {
```

Comment:
Is there a way we can do this using native Promises instead of introducing a new promise library?

Reply:
using bluebird mainly for the concurrency limit, to avoid trying to require all files immediately